[FEAT] 관리자 자격증명 페이지네이션 + 필터링 도입

kimdevv opened the pull request:
## Issue Number
closed #886 

## As-Is
* 자격증명 데이터의 양이 늘어나 관리자 페이지에서 전체 자격증명 조회에 걸리는 시간이 증가하였습니다.

## To-Be
* 관리자 전용 자격증명 조회 로직을 기존 CertificateService에서 관리자 전용 서비스 레이어로 분리하였습니다.
* 관리자 페이지의 자격증명 조회 방식을 페이지 조회로 변경하였습니다.
* 관리자 자격증명 조회 시 상태로 필터링하는 로직을 일부 수정하였습니다.
* QueryDSL을 사용하여 이루어졌습니다.

## Check List
- [x] 테스트가 전부 통과되었나요?
- [x] 모든 commit이 push 되었나요?
- [x] merge할 branch를 확인했나요?
- [x] Assignee를 지정했나요?
- [x] Label을 지정했나요?
- [x] 닫을 이슈 번호를 지정했나요?

## (Optional) Additional Description
[관리자 멤버 목록 페이지네이션 도입 브랜치](https://github.com/woowacourse-teams/2025-Fit-toring/pull/881)를 참고하여 거의 같은 로직으로 구현하였습니다.

kimdevv committed:
refactor: 관리자 자격증명 서비스 분리 및 리팩토링

- CertificateService에서 관리자 관련 로직 분리하여 AdminCertificateService 생성
- 관련 컨트롤러 및 테스트 클래스 수정
- MemberRepository 및 필요 없는 메서드 제거로 코드 간소화
- Docker 포트 설정 수정 및 비활성화된 서비스 주석 처리
- 관리자의 자격증명 조회, 승인, 거절 로직 단순화 및 코드 구조화

kimdevv committed:
feat: 관리자 자격증명 페이징 및 필터링 조회 기능 추가

- findAllWithFilterAndPagination 메서드로 필터링 및 페이징 처리 구현
- countByStatus 메서드로 상태별 데이터 개수 조회 기능 추가
- AdminCertificateResponse 및 PageResult DTO 생성
- CertificateRepository에 CustomCertificateRepository 인터페이스 확장
- 서비스 및 컨트롤러 계층에 페이징 로직 통합 및 기존 로직 대체

kimdevv committed:
test: 관리자 자격증명 페이징 및 필터링 테스트 추가

- getAllCertificatesPaged 메서드로 페이지네이션 테스트 추가
- 상태별 필터링 후 데이터 검증을 위한 테스트 구현
- 테스트 데이터 생성 로직 개선 및 코드 간소화

jye326 commented on backend/fittoring/src/main/java/fittoring/admin/repository/CustomCertificateRepositoryImpl.java:
```diff
@@ -0,0 +1,54 @@
+        long offset = (long) (page - 1) * size;
+
+        return jpaQueryFactory.select(Projections.constructor(AdminCertificateResponse.class,
+                certificate.id, certificate.mentoring.mentor.name, certificate.title, certificate.type, certificate.verificationStatus, certificate.createdAt)
```
p3: 프로젝션 생성자도 개행해 볼까요?

jye326 commented on backend/fittoring/src/main/java/fittoring/admin/repository/CustomCertificateRepositoryImpl.java:
```diff
@@ -0,0 +1,54 @@
+    private final JPAQueryFactory jpaQueryFactory;
+
+    @Override
+    public List<AdminCertificateResponse> findAllWithFilterAndPagination(Status status, int page, int size) {
```
p2: offset 페이징은 앞의 offset만큼을 읽고 버립니다. 버리는 칼럼에 대한 IO를 최소화 하기 위해 필요한 Id만 먼저 조회하고, 해당 id의 칼럼값을 채우는 방법을 사용할 수 있을 것 같아요!

jye326 commented on backend/fittoring/src/test/java/fittoring/admin/service/AdminCertificateServiceTest.java:
```diff
@@ -0,0 +1,304 @@
+            softAssertions.assertThat(sixthResponse.total()).isEqualTo(105);
+            softAssertions.assertThat(sixthResponse.totalPages()).isEqualTo(6);
+            softAssertions.assertThat(sixthResponse.page()).isEqualByComparingTo(6);
+            softAssertions.assertThat(sixthResponse.hasNext()).isFalse();
```
😮

ppparkta replied:
엄청나네요.

jye326 reviewed (approved):
수고하셨습니다! 코멘트 남기고 승인할게요!

ppparkta commented on backend/fittoring/src/main/java/fittoring/admin/presentation/dto/PageResult.java:
```diff
@@ -0,0 +1,13 @@
+    int size,
+    long total,
+    int totalPages,
+    boolean hasNext
```
p1: 사용하지 않는 필드는 제거해주세요. hasNext는 무한스크롤 방식에서 쓰였는데 지금 통일된 방법에서는 쓰이지 않을 것 같아요.

ppparkta commented on backend/fittoring/src/main/java/fittoring/admin/presentation/dto/PageResult.java:
```diff
@@ -0,0 +1,13 @@
+    int totalPages,
+    boolean hasNext
+) {
+}
```
p3: 개행 추가해주세요!

ppparkta commented on backend/fittoring/src/main/java/fittoring/admin/repository/CustomCertificateRepositoryImpl.java:
```diff
@@ -0,0 +1,54 @@
+
+    private static final QCertificate certificate = QCertificate.certificate;
+    private static final QMentoring mentoring = QMentoring.mentoring;
+    private static final QMember member = QMember.member;
```
p2: 상수 컨벤션에 맞춰 대문자로 수정해주세요! 더불어 일반 필드와의 구분을 위해 개행이 들어가면 좋을 것 같습니다.

ppparkta commented on backend/fittoring/src/main/java/fittoring/admin/service/AdminCertificateService.java:
```diff
@@ -0,0 +1,81 @@
+
+    public PageResult<AdminCertificateResponse> getAllCertificatesPaged(Long memberId, Status status, int page) {
+        checkAdminAuthority(memberId);
+        List<AdminCertificateResponse> certificates = certificateRepository.findAllWithFilterAndPagination(status, page, PAGE_SIZE_OF_CERTIFICATE);
```
p3: 매개변수에 개행이 적용되면 좋을 것 같아요!

ppparkta commented on backend/fittoring/src/main/java/fittoring/admin/service/AdminCertificateService.java:
```diff
@@ -0,0 +1,81 @@
+@Service
+public class AdminCertificateService {
+
+    public static final int PAGE_SIZE_OF_CERTIFICATE = 20;
```
p3: 여기도 마찬가지로 개행 한 번 들어가면 구분되어 더 보기 편할 것 같아요

ppparkta commented on backend/fittoring/src/main/java/fittoring/admin/service/AdminCertificateService.java:
```diff
@@ -0,0 +1,81 @@
+    private final CertificateRepository certificateRepository;
+    private final MemberRepository memberRepository;
+
+    public PageResult<AdminCertificateResponse> getAllCertificatesPaged(Long memberId, Status status, int page) {
```
p1: 트랜잭션이 누락되었네요. `@Transactional(readOnly=true)` 부탁드려요!

ppparkta commented on backend/fittoring/src/main/java/fittoring/admin/service/AdminCertificateService.java:
```diff
@@ -0,0 +1,81 @@
+        }
+    }
+
+    public CertificateDetailResponse getCertificate(Long memberId, Long certificateId) {
```
p1: 트랜잭션이 누락되었네요. `@Transactional(readOnly=true)` 부탁드려요!

ppparkta reviewed (approved):
고생하셨습니다. 코멘트 몇 개 남겨두었습니다!

kimdevv committed:
Merge branch 'develop' into feature/886-admin-certificate

# Conflicts:
#	backend/fittoring/src/main/java/fittoring/admin/presentation/dto/PageResult.java
#	backend/fittoring/src/main/java/fittoring/application/mentoring/repository/CertificateRepository.java
#	backend/fittoring/src/main/java/fittoring/application/mentoring/service/CertificateService.java

kimdevv committed:
feat: 관리자 회원 조회 기능에 트랜잭션 추가

- findAllForAdminPaged 메서드에 @Transactional(readOnly=true) 어노테이션 추가
- 관리자 회원 조회 시 데이터 정합성 보장을 위한 설정 적용

kimdevv committed:
refactor: PageResult의 hasNext 필드 제거

- hasNext 필드를 더 이상 사용하지 않아 관련 코드 삭제
- 불필요한 로직 간소화 및 가독성 개선

kimdevv committed:
refactor: 상수 네이밍 변경 및 쿼리 가독성 개선

- 상수명을 대문자 스네이크 케이스로 변경
- JPAQueryFactory에서 사용되는 쿼리 가독성을 위해 상수 활용
- 코드 정렬 및 불필요한 공백 제거로 유지보수성 향상

kimdevv committed:
feat: 관리자 자격증명 조회 메서드에 트랜잭션 추가

- getAllCertificatesPaged, getCertificate 메서드에 @Transactional(readOnly=true) 어노테이션 추가
- 데이터 조회 시 정합성 보장을 위해 트랜잭션 적용

kimdevv committed:
refactor: 관리자 자격증명 조회 로직 분리 및 쿼리 수정

- findAllWithFilterAndPagination 메서드를 findCertificateIdsForAdmin과 findCertificatesByIdsOrdered로 분리
- ID 조회와 엔티티 조회를 분리해 쿼리 성능 최적화
- AdminCertificateService에 변경된 메서드 로직 반영
- 코드 가독성 및 유지보수성 개선

kimdevv committed:
feat: 관리자 자격증명 조회 메서드에 페이지 크기 동적 적용

- getAllCertificatesPaged 메서드에 페이지 크기(size)를 인자로 추가하여 동적으로 설정 가능하도록 수정
- 상수 값(PAGE_SIZE_OF_CERTIFICATE) 제거 및 관련 테스트 코드 수정
- 페이지 크기 변경에 따른 totalPages 계산 로직 업데이트
- 관련 서비스, 컨트롤러, 테스트 클래스 로직 수정 및 동작 검증

kimdevv committed:
refactor: CertificateService에서 관리자 전용 로직 제거

- 관리자 관련 메서드(getAllCertificates, approveCertificate 등) 삭제
- CertificateService의 역할을 일반 사용자 관련 로직으로 제한
- 불필요한 관리자 권한 체크 및 예외 처리 코드 제거
- 코드 간소화로 가독성 및 유지보수성 향상

kimdevv committed:
Merge branch 'develop' into feature/886-admin-certificate

# Conflicts:
#	backend/fittoring/src/main/java/fittoring/application/mentoring/service/CertificateService.java
#	backend/fittoring/src/test/java/fittoring/application/mentoring/service/CertificateServiceTest.java
